openhab/openhab2-addons
# HueEmulation: Fixes and test bundle

Unfortunately PR #4313 broke the HueEmulation (within the class hierarchy of `hueState` is two times a field with the name "alert" and that cannot be serialised by gJson). I have therefore created a Test bundle for the HueEmulation to catch those breaking commits.

Refactor
* Create subpackages in internal.dto
* Move REST API parts out of HueEmulationServlet into RestApi class (two classes < 300 lines of code instead of one with > 600 lines)
* Don't let HueEmulationServlet extend HttpServlet, but use an anonymous class field.
  This is because two servlets are required.
* Rename HueEmulationServlet to HueEmulationService
* Hierarchical processing of the REST URI in RestApi class
  for easier maintenance and extendability.
* Use StorageService instead of files for storing users, light mapping and
  user modified bridge parameters.

Rest API Extensions:
* Support POST on /api/{username}/lights
* Support GET on  /api/{username}/lights/new
* Support PUT on  /api/{username}/config (for renaming the emulated bridge)
* Support for groups on /api/{username}/groups

Support for /description.xml next to /api/description.xml, because some Philips devices expect a Hue bridge to have that file on the root path, no matter what the upnp discovery message says.

Test bundle:
* Full implementation test with
  - check for /description.xml upnp file
  - check for invalid user access, valid user access, user creation
  - access for /api/{username}/lights and /api/{username}/lights/{id}
* Unit tests for RestApi class for
  - user management
  - request lights, change light state

Signed-off-by: David Graeff <david.graeff@web.de>
